frontend/src/metabase/dashboard/context/context.tsx

```diff
@@ -121,69 +125,85 @@ const DashboardContextProviderInner = ({
   const previousDashboard = usePrevious(dashboard);
   const previousDashboardId = usePrevious(dashboardId);
   const previousTabId = usePrevious(selectedTabId);
   const previousParameterValues = usePrevious(parameterValues);
 
   const shouldRenderAsNightMode = Boolean(isNightMode && isFullscreen);
 
-  const handleLoadDashboard = useCallback(
-    async (dashboardId: DashboardId) => {
-      initialize({ clearCache: !isNavigatingBackToDashboard });
-
-      const result = await fetchDashboard({
-        dashId: dashboardId,
-        queryParams: parameterQueryParams,
-        options: {
-          clearCache: !isNavigatingBackToDashboard,
-          preserveParameters: isNavigatingBackToDashboard,
-        },
-      });
-
-      return result;
-    },
-    [
-      fetchDashboard,
-      initialize,
-      isNavigatingBackToDashboard,
-      parameterQueryParams,
-    ],
-  );
-
   const handleError = useCallback(
-    (err: FailedFetchDashboardResult | Error) => {
-      const error = err instanceof Error ? { error: err, payload: err } : err;
+    (error: Error) => {
       onError?.(error);
       setError(error);
     },
     [onError],
   );
 
-  useEffect(() => {
-    const hasDashboardChanged = dashboardId !== previousDashboardId;
-    if (hasDashboardChanged) {
-      setError(null);
-      handleLoadDashboard(dashboardId)
-        .then((result) => {
-          if (isFailedFetchDashboardResult(result)) {
-            handleError(result);
-          }
-        })
-        .catch((err) => {
-          handleError(err);
-        });
-      return;
-    }
+  const fetchId = useCallback(
+    async (idToFetch: DashboardId) => {
+      try {
+        const { id, isError } = await dispatch(
+          fetchEntityId({ type: "dashboard", id: idToFetch }),
+        );
+        if (isError || id === null) {
+          handleError({
+            status: 404,
+            message: "Not found",
+            name: "Not found",
+          } as Error);
+          setDashboardId(id);
+        }
+        setDashboardId(id);
+      } catch (e) {
+        handleError(e as Error);
+      }
 
-    if (!dashboard) {
       return;
-    }
+    },
+    [dispatch, handleError],
+  );
+
+  const fetchData = useCallback(
+    async (dashboardId: DashboardId) => {
+      const hasDashboardChanged = dashboardId !== previousDashboardId;
+      if (hasDashboardChanged) {
+        setError(null);
+
+        initialize({ clearCache: !isNavigatingBackToDashboard });
+        fetchDashboard({
+          dashId: dashboardId,
+          queryParams: parameterQueryParams,
+          options: {
+            clearCache: !isNavigatingBackToDashboard,
+            preserveParameters: isNavigatingBackToDashboard,
+          },
+        })
+          .then((result) => {
+            if (isFailedFetchDashboardResult(result)) {
+              handleError(result.payload as Error);
+            }
+          })
+          .catch((err) => handleError(err as Error));
```

Comment:
I see a lot of type casts for `handleError`, should we change its signature to accept `unknown` instead?